ethereum/core/src/main/java/tech/pegasys/teku/core/stategenerator/StateRegenerationBaseSelector.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.core.stategenerator;
+
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import tech.pegasys.teku.core.lookup.BlockProvider;
+import tech.pegasys.teku.core.lookup.StateAndBlockProvider;
+import tech.pegasys.teku.datastructures.blocks.SignedBlockAndState;
+import tech.pegasys.teku.datastructures.blocks.SlotAndBlockRoot;
+import tech.pegasys.teku.datastructures.state.BlockRootAndState;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+
+public class StateRegenerationBaseSelector {
+  private final Optional<SlotAndBlockRoot> latestEpochBoundary;
+  private final Supplier<Optional<BlockRootAndState>> closestAvailableStateSupplier;
+  private final StateAndBlockProvider stateAndBlockProvider;
+  private final BlockProvider blockProvider;
+  private final Optional<SignedBlockAndState> rebasedStartingPoint;
```

Comment:
Probably worth trying this as-is, but I wonder if we should stop holding onto the rebased starting point as well?  Presumably this should be in the `Store` when the task runs, right?

Reply:
Yeah I went back and forth on that.  I'm worried that if we don't hold onto it we'll wind up constantly regenerating the same states because the store only holds them with soft references. The state generation queue rebases new tasks onto existing ones but doesn't otherwise prioritise regeneration tasks to process all of a single fork first for example and it doesn't detect when existing tasks could be rebased onto new ones.

I suspect that will need to be the next step - especially since that's where all the memory usage comes from for checkpoint states.